pkg/exchange/coinbase/exchage.go

```diff
@@ -0,0 +1,309 @@
+package coinbase
+
+import (
+	"context"
+	"strings"
+	"time"
+
+	api "github.com/c9s/bbgo/pkg/exchange/coinbase/api/v1"
+	"github.com/c9s/bbgo/pkg/types"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	"golang.org/x/time/rate"
+)
+
+var (
+	// Rate Limit: 10 requests per 2 seconds, Rate limit rule: UserID
+	queryAccountLimiter = rate.NewLimiter(rate.Every(200*time.Millisecond), 1)
+
+	// compile time check: implemented interface
+	_ types.Exchange                  = &Exchange{}
+	_ types.ExchangeMarketDataService = &Exchange{}
+	_ types.CustomIntervalProvider    = &Exchange{}
+)
+
+const (
+	ID            = "coinbase"
+	PlatformToken = "COIN"
+)
+
+var log = logrus.WithField("exchange", ID)
+
+type Exchange struct {
+	client *api.RestAPIClient
+}
+
+func New(key, secret, passphrase string, timeout time.Duration) *Exchange {
+	client := api.NewClient(key, secret, passphrase, timeout)
+	return &Exchange{
+		client: client,
+	}
+}
+
+// CustomIntervalProvider
+var supportedIntervalMap = map[types.Interval]int{
+	types.Interval1m:  60,
+	types.Interval5m:  5 * 60,
+	types.Interval15m: 15 * 60,
+	types.Interval1h:  60 * 60,
+	types.Interval6h:  6 * 60 * 60,
+	types.Interval1d:  24 * 60 * 60,
+}
+
+func (e *Exchange) SupportedInterval() map[types.Interval]int {
+	return supportedIntervalMap
+}
+
+func (e *Exchange) IsSupportedInterval(interval types.Interval) bool {
+	_, ok := supportedIntervalMap[interval]
+	return ok
+}
+
+// ExchangeMinimal
+func (e *Exchange) Name() types.ExchangeName {
+	return types.ExchangeCoinBase
+}
+
+func (e *Exchange) PlatformFeeCurrency() string {
+	return PlatformToken
+}
+
+// ExchangeAccountService
+func (e *Exchange) QueryAccount(ctx context.Context) (*types.Account, error) {
+	if err := queryAccountLimiter.Wait(ctx); err != nil {
+		return nil, err
+	}
+
+	balances, err := e.QueryAccountBalances(ctx)
+	if err != nil {
+		return nil, err
+	}
+	account := types.NewAccount()
+	account.UpdateBalances(balances)
+	return account, nil
+}
+
+func (e *Exchange) QueryAccountBalances(ctx context.Context) (types.BalanceMap, error) {
+	if err := queryAccountLimiter.Wait(ctx); err != nil {
+		return nil, err
+	}
+
+	req := e.client.NewGetBalancesRequest()
+	accounts, err := req.Do(ctx)
+	if err != nil {
+		return nil, err
+	}
+	balances := make(types.BalanceMap)
+	for _, cbBalance := range accounts {
+		cur := strings.ToUpper(cbBalance.Currency)
+		balances[cur] = *toGlobalBalance(cur, &cbBalance)
+	}
+	return balances, nil
+}
+
+// ExchangeTradeService
+func (e *Exchange) SubmitOrder(ctx context.Context, order types.SubmitOrder) (createdOrder *types.Order, err error) {
+	req := e.client.NewCreateOrderRequest().OrderType(string(order.Type)).Side(string(order.Side)).ProductID(toLocalSymbol(order.Symbol)).Size(order.Quantity.String())
+	res, err := req.Do(ctx)
+	if err != nil {
+		return nil, err
+	}
+	return &types.Order{
+		Exchange: types.ExchangeCoinBase,
+		UUID:     res.ID,
```

Comment:
you need to assign SubmitOrder into the SubmitOrder field,

and you might need to query the order details, take a look at bybit or bitget's SubmitOrder() implementation

Reply:
Got it.

Reply:
Done.